### ISSUE TITLE
systemd-run --pty overflows terminal title with GNU screen, causing it to dump part of it to the terminal

### ISSUE THREAD

**Issue by mk-fg** [systemd/systemd#35104]
### systemd version the issue has been seen with

256.7

### Used distribution

Arch Linux

### Linux kernel version used

6.6.57-desktop-v4

### CPU architectures issue was seen on

x86_64

### Component

systemd-run

### Expected behaviour you didn't see

1. Running `systemd-run -q --user --pty -- true ...<1000 characters of arguments>` produces empty terminal output, same as "true" command.
2. Running `systemd-run -q --user --pty -- true ...` with command like "true" that never produces any output consistently doesn't produce any output either, regardless of arguments to that command, unless there's some issue with being able to run it (e.g. command-line too long).

### Unexpected behaviour you saw

Running `systemd-run -q --user --pty -- true A B C` works silently as expected.

But running this long command (>1000 characters are generated) seem to dump contents of some internal buffer back to stdout (maybe starting at offset 750 if it contains `<arguments> on <hostname>`):
``` console
% systemd-run -q --user --pty -- true {1..300}
5 216 217 218 219 ... 300 on myhostname
```

`--pty` option seem to be required to reproduce this behavior.
`-q --user` are added to run this easier as unprivileged user and get cleaner output.
`{1..300}` shell syntax prints space-separated numbers from 1 to 300, which are used as arguments for "true".
Bumped into it when running long qemu commands wrapped into systemd-run.

### Steps to reproduce the problem

Run `systemd-run -q --user --pty -- true {1..300}` command with systemd 256.7.
Expected result is no output (as there is none from any `true ...` invocation), actual result is a long line of `arg... on <hostname>` from systemd-run.

**Comment by mk-fg** [comment]
Had a thought to try this on an system with earlier systemd, and found that while this behavior is present with `systemd 256 (256.7-1-arch)`, it's not there with `systemd 256 (256.4-1-arch)`, so might have been introduced relatively recently.

**Comment by poettering** [comment]
what terminal are you using?

my educated guess: we nowadays set the window title to the cmdline. And I guess if it gets overly long your terminal will truncate it and then assume the rest of the string is just regular text again.

**Comment by poettering** [comment]
if you invoke your command with the env var SYSTEMD_ADJUST_TERMINAL_TITLE=0 set, does the issue go away?

**Comment by mk-fg** [comment]
Yes, indeed setting terminal title seem to be the issue here.

- `SYSTEMD_ADJUST_TERMINAL_TITLE=0 systemd-run -q --user --pty -- true {1..300}` does not dump cmdline anywhere.
- `systemd-run -q --user --pty -- true {1..300}` dumps cmdline in `screen` (as in [GNU Screen](https://www.gnu.org/software/screen/)) running in xterm.
- `systemd-run -q --user --pty -- true {1..300}` does not dump cmdline in `tmux` or plain xterm.
- Previous mention of systemd 256.7 vs 256.4 version affecting the issue was bogus, running with `256.4` in `screen` has same issue.

So it looks like GNU screen has this problem with an overly long title, and not tmux/xterm, at least from this limited testing.

**Comment by mk-fg** [comment]
Tried running it with two recent screen releases:
```
Screen version 4.09.01 (GNU) 20-Aug-23
Screen version 5.0.0 (build on 2024-08-29 01:08:49)
```
Both seem this issue, and I guess it's a bug to report there.

But also maybe systemd can truncate the title as well? Seems unlikely that it'd be useful at 1k characters anyway.

**Comment by mk-fg** [comment]
> Can you try with SYSTEMD_LOG_LEVEL=debug systemd-run ... to see if there are any clues wrt output?

In case this might still be useful as well:

```
# SYSTEMD_LOG_LEVEL=debug systemd-run -q --user --pty -- true {1..300}
Bus n/a: changing state UNSET → OPENING
sd-bus: starting bus by connecting to /run/user/0/bus...
Bus n/a: changing state OPENING → AUTHENTICATING
Bus n/a: changing state AUTHENTICATING → HELLO
Sent message type=method_call sender=n/a destination=org.freedesktop.DBus path=/org/freedesktop/DBus interface=org.freedesktop.DBus member=Hello cookie=1 reply_cookie=0 signature=n/a error-name=n/a error-message=n/a
Got message type=method_return sender=org.freedesktop.DBus destination=:1.32 path=n/a interface=n/a member=n/a  cookie=4294967295 reply_cookie=1 signature=s error-name=n/a error-message=n/a
Bus n/a: changing state HELLO → RUNNING
Sent message type=method_call sender=n/a destination=org.freedesktop.systemd1 path=/org/freedesktop/systemd1 interface=org.freedesktop.systemd1.Manager member=StartTransientUnit cookie=2 reply_cookie=0 signature=ssa(sv)a(sa(sv)) error-name=n/a error-message=n/a
Got message type=method_return sender=:1.1 destination=:1.32 path=n/a interface=n/a member=n/a  cookie=1434 reply_cookie=2 signature=o error-name=n/a error-message=n/a
Sent message type=method_call sender=n/a destination=org.freedesktop.DBus path=/org/freedesktop/DBus interface=org.freedesktop.DBus member=AddMatch cookie=3 reply_cookie=0 signature=s error-name=n/a error-message=n/a
Sent message type=method_call sender=n/a destination=org.freedesktop.systemd1 path=/org/freedesktop/systemd1/unit/run_2du32_2eservice interface=org.freedesktop.DBus.Properties member=GetAll cookie=4 reply_cookie=0 signature=s error-name=n/a error-message=n/a
Got message type=method_return sender=:1.1 destination=:1.32 path=n/a interface=n/a member=n/a  cookie=1442 reply_cookie=4 signature=a{sv} error-name=n/a error-message=n/a
5 216 217 218 219 220 221 222 223 224 225 226 227 228 229 230 231 232 233 234 235 236 237 238 239 240 241 242 243 244 245 246 247 248 249 250 251 252 253 254 255 256 257 258 259 260 261 262 263 264 265 266 267 268 269 270 271 272 273 274 275 276 277 278 279 280 281 282 283 284 285 286 287 288 289 290 291 292 293 294 295 296 297 298 299 300 on testGot message type=signal sender=org.freedesktop.DBus destination=:1.32 path=/org/freedesktop/DBus interface=org.freedesktop.DBus member=NameAcquired  cookie=4294967295 reply_cookie=0 signature=s error-name=n/a error-message=n/a
Got message type=method_return sender=org.freedesktop.DBus destination=:1.32 path=n/a interface=n/a member=n/a  cookie=4294967295 reply_cookie=3 signature= error-name=n/a error-message=n/a
Match type='signal',sender='org.freedesktop.systemd1',path='/org/freedesktop/systemd1/unit/run_2du32_2eservice',interface='org.freedesktop.DBus.Properties',member='PropertiesChanged' successfully installed.
Got message type=signal sender=:1.1 destination=n/a path=/org/freedesktop/systemd1/unit/run_2du32_2eservice interface=org.freedesktop.DBus.Properties member=PropertiesChanged  cookie=1443 reply_cookie=0 signature=sa{sv}as error-name=n/a error-message=n/a
Sent message type=method_call sender=n/a destination=org.freedesktop.systemd1 path=/org/freedesktop/systemd1/unit/run_2du32_2eservice interface=org.freedesktop.DBus.Properties member=GetAll cookie=5 reply_cookie=0 signature=s error-name=n/a error-message=n/a
Got message type=method_return sender=:1.1 destination=:1.32 path=n/a interface=n/a member=n/a  cookie=1451 reply_cookie=5 signature=a{sv} error-name=n/a error-message=n/a
Got message type=signal sender=:1.1 destination=n/a path=/org/freedesktop/systemd1/unit/run_2du32_2eservice interface=org.freedesktop.DBus.Properties member=PropertiesChanged  cookie=1444 reply_cookie=0 signature=sa{sv}as error-name=n/a error-message=n/a
Sent message type=method_call sender=n/a destination=org.freedesktop.systemd1 path=/org/freedesktop/systemd1/unit/run_2du32_2eservice interface=org.freedesktop.DBus.Properties member=GetAll cookie=6 reply_cookie=0 signature=s error-name=n/a error-message=n/a
Got message type=method_return sender=:1.1 destination=:1.32 path=n/a interface=n/a member=n/a  cookie=1452 reply_cookie=6 signature=a{sv} error-name=n/a error-message=n/a
Got message type=signal sender=:1.1 destination=n/a path=/org/freedesktop/systemd1/unit/run_2du32_2eservice interface=org.freedesktop.DBus.Properties member=PropertiesChanged  cookie=1445 reply_cookie=0 signature=sa{sv}as error-name=n/a error-message=n/a
Sent message type=method_call sender=n/a destination=org.freedesktop.systemd1 path=/org/freedesktop/systemd1/unit/run_2du32_2eservice interface=org.freedesktop.DBus.Properties member=GetAll cookie=7 reply_cookie=0 signature=s error-name=n/a error-message=n/a
Got message type=method_return sender=:1.1 destination=:1.32 path=n/a interface=n/a member=n/a  cookie=1453 reply_cookie=7 signature=a{sv} error-name=n/a error-message=n/a
Got message type=signal sender=:1.1 destination=n/a path=/org/freedesktop/systemd1/unit/run_2du32_2eservice interface=org.freedesktop.DBus.Properties member=PropertiesChanged  cookie=1446 reply_cookie=0 signature=sa{sv}as error-name=n/a error-message=n/a
Sent message type=method_call sender=n/a destination=org.freedesktop.systemd1 path=/org/freedesktop/systemd1/unit/run_2du32_2eservice interface=org.freedesktop.DBus.Properties member=GetAll cookie=8 reply_cookie=0 signature=s error-name=n/a error-message=n/a
Got message type=method_return sender=:1.1 destination=:1.32 path=n/a interface=n/a member=n/a  cookie=1458 reply_cookie=8 signature=a{sv} error-name=n/a error-message=n/a
Bus n/a: changing state RUNNING → CLOSED
```

This is output copied from a terminal, so includes what seem to be a title overflow bits.
Don't think anything looks odd in that debug output.

**Comment by mk-fg** [comment]
Thanks for the fix/workaround, didn't remember about these vars at all unfortunately.

Issue does not seem to be with systemd, but dunno if maybe it'd be useful to truncate long titles for better compatibility in systemd-run as well, and whether it might be useful to keep open for that, feel free to close otherwise.

**Comment by poettering** [comment]
Prepped a fix in #35142.

I think screen is actually doing the right thing, the should put a limit on the length of ansi sequences they accept.

So this is our bug, not theirs. 

Would be good though to know where precisely screen (and the other relevant emulators) put the limit though. For now I ellipsize at 128, but we could probably increase it if that's safe.
